relay_sidecar/save_ops.py

```
@@ -0,0 +1,95 @@
+
+    def __init__(self, workspace_root: Path, *, backup_extension: str = ".bak") -> None:
+        self.workspace_root = workspace_root.expanduser().resolve()
+        self.backup_extension = backup_extension
+        self.workspace_root.mkdir(parents=True, exist_ok=True)
+
+    # Path helpers ------------------------------------------------------
+
+    def resolve_workspace_path(self, path: Path) -> Path:
+        candidate = (self.workspace_root / path).expanduser().resolve()
+        if not str(candidate).startswith(str(self.workspace_root)):
+            raise SaveError(f"path {candidate} escapes workspace root {self.workspace_root}")
```

Ruff (0.13.1)

Avoid specifying long messages outside the exception class

(TRY003)